lib/protocol/protocol.go

```diff
@@ -938,9 +960,13 @@ func (c *rawConnection) internalClose(err error) {
 		}
 		c.awaitingMut.Unlock()
 
-		<-c.dispatcherLoopStopped
+		if !c.startTime.IsZero() {
+			// Wait for the dispatcher loop to exit, if it was started to
+			// begin with.
+			<-c.dispatcherLoopStopped
+		}
```

Comment:
I guess that caused issues during testing/debugging? I'd put that into a separate commit/PR then, just so we don't lose this fix if for some reason this entire refactoring gets reverted.